front/lib/api/assistant/conversation.ts

```diff
@@ -630,16 +580,134 @@ export async function* retryAgentMessage(
+    agentMessageRow: AgentMessage;
+  } | null = await front_sequelize.transaction(async (t) => {
+    const messageRow = await Message.findOne({
+      where: {
+        conversationId: conversation.id,
+        id: message.id,
+      },
+      include: [
+        {
+          model: AgentMessage,
+          as: "agentMessage",
+          required: true,
+        },
+      ],
+      transaction: t,
+    });
+
+    if (!messageRow || !messageRow.agentMessage) {
+      return null;
+    }
+    const agentMessageRow = messageRow.agentMessage;
+    const m = await Message.create(
+      {
+        sId: messageRow.sId,
+        rank: messageRow.rank,
+        conversationId: conversation.id,
+        parentId: messageRow.parentId,
+        version: messageRow.version + 1,
+        agentMessageId: (
+          await AgentMessage.create(
+            {
+              status: "created",
+              agentConfigurationId: agentMessageRow.agentConfigurationId,
+            },
+            { transaction: t }
+          )
+        ).id,
+      },
+      {
+        transaction: t,
+      }
+    );
+    const agentMessage: AgentMessageType = {
+      id: m.id,
+      sId: m.sId,
+      type: "agent_message",
+      visibility: m.visibility,
+      version: m.version,
+      parentMessageId: message.parentMessageId,
+      status: "created",
+      action: null,
+      content: null,
+      feedbacks: [],
+      error: null,
+      configuration: message.configuration,
+    };
+    return {
+      agentMessage,
+      agentMessageRow,
+    };
+  });
+
+  if (!agentMessageResult) {
+    return {
+      type: "agent_error",
+      created: Date.now(),
+      configurationId: message.configuration.sId,
+      messageId: message.sId,
+      error: {
+        code: "message_not_found",
+        message: "The message to retry was not found",
+      },
+    };
+  }
+
+  const { agentMessage, agentMessageRow } = agentMessageResult;
+
   yield {
-    type: "agent_error",
+    type: "agent_message_new",
     created: Date.now(),
-    configurationId: "foo",
-    messageId: "bar",
-    error: {
-      code: "not_implemented",
-      message: "Not implemented",
-    },
+    configurationId: agentMessage.configuration.sId,
+    messageId: agentMessage.sId,
+    message: agentMessage,
   };
+
+  // We stitch the conversation to retry the agent message correctly: no other
+  // messages than this agent's past its parent message.
+
+  // First, find the array of the parent message in conversation.content.
+  const parentMessageIndex = conversation.content.findIndex((messages) => {
+    return messages.some((m) => m.sId === message.parentMessageId);
+  });
+
+  // Then, find this agentmessage's array in conversation.content and add the
+  // new agent message to it.
+  const agentMessageArray = conversation.content.find((messages) => {
+    return messages.some((m) => m.sId === message.sId && isAgentMessageType(m));
+  }) as AgentMessageType[];
```

Comment:
cast seems necessary here, not sure how to get typescript how to understand it otherwise

furthermore feels legit since nonexistence would be PLS and would immediately throw the line below, and assistantmessagetype is checked inside

Reply:
yep this type is hard to work with